Neos.ContentRepository/Classes/Domain/Service/NodeMoveIntegrityCheckService.php

```diff
@@ -0,0 +1,161 @@
+<?php
+
+namespace Neos\ContentRepository\Domain\Service;
+
+/*
+ * This file is part of the Neos.ContentRepository package.
+ *
+ * (c) Contributors of the Neos Project - www.neos.io
+ *
+ * This package is Open Source Software. For the full copyright and license
+ * information, please view the LICENSE file which was distributed with this
+ * source code.
+ */
+
+use Neos\ContentRepository\Domain\Model\Node;
+use Neos\ContentRepository\Domain\Model\NodeInterface;
+use Neos\ContentRepository\Domain\Service\Dto\NodeMoveIntegrityCheckResult;
+use Neos\ContentRepository\Domain\Service\Dto\NodeMoveIntegrityCheckResultPart;
+use Neos\ContentRepository\Domain\Utility\NodePaths;
+use Neos\ContentRepository\Exception\NodeMoveIntegrityViolationException;
+use Neos\Flow\Annotations as Flow;
+
+/**
+ * This is an internal service to prevent disconnected nodes when moving.
+ *
+ * It is called from {@see Node::setPathInternalForAggregate()}, and throws an error in case disconnected nodes might appear
+ * due to a move.
+ *
+ * ## How can disconnected nodes occur?
+ *
+ * Background: When moving a node, it is moved across ALL dimensions.
+ *
+ * (This issue is described at https://github.com/neos/neos-development-collection/issues/3384 )
+ *
+ * Let's say we have the following Node structure in the beginning:
+ *
+ * |-- site (de, en)
+ * . |-- cr  (de, en)
+ * . | |-- subpage (de, en)
+ * . |-- other  (de)
+ *
+ * Now, the user moves /site/cr/subpage underneath /site/other/ in the user workspace.
+ *
+ * Until October 2021, this lead to the following situation / bug:
```

Comment:
```suggestion
 * Until this was merged, this lead to the following situation / bug:
```

Otherwise this will eventually confuse the hell out of someone in the future…